Litmus concepts (#3)
puppetlabs/litmus: litmus-concepts.md

  * hypervisors: docker, vagrant, vmpooler, abs
  * external provisioners: e.g. terraform
  * test systems:
    * litmusimage
    * upstream images
    * custom images
* utility code
  * puppetlabs-facts module
  * puppetlabs-puppet_agent module
* Test Tracing: honeycomb
* Testing System
  * runner: RSpec
  * test case definition: RSpec
  * test setup:
    * a manifest string embedded in the test case
    * ruby code to orchestrate a change
  * test assertion
    * serverspec
    * hand-coded checks
* Packaging and delivery:
  * Litmus as gem
  * Litmus as PDK component
  * utility modules as git repos
  * bolt as gem

The following sections go over the various components, their reasoning and discuss alternative options.

## UI

The current UI/UX of Litmus is implemented as a set of rake tasks in [puppet_litmus:lib/puppet_litmus/rake_tasks.rb](https://github.com/puppetlabs/puppet_litmus/blob/master/lib/puppet_litmus/rake_tasks.rb).

**Reasoning:**
Rake is a ubiquitous choice of interacting with a set of connected tasks in the ruby world.
The limited option parsing capabilities push the design towards simple interactions and storing important information in configuration files.

> It is very easy to get started with a project based on rake tasks, as there is a wealth of examples and titorials how to build rake tasks.

Reviewer: ```suggestion
It is very easy to get started with a project based on rake tasks, as there is a wealth of examples and tutorials on how to build rake tasks.
```

Author: oops! :shame:

This could be implemented as part of the pdk (`pdk test acceptance`) or a standalone CLI tool shipped as part of the PDK.

The VSCode plugin could provide additional UI to make running and interacting with acceptance tests directly from the IDE possible.

## Communication Layer

The communications layer for litmus needs to be able to talk to all the various systems that users might want to use in testing.

**Reasoning:**
Bolt supports SSH, WinRM natively and allows extension to other remote protocols using Puppet's extension points.
With Bolt being integrated into our entire product line, re-using it also for content testing is an easy choice.
Content generated for Bolt can be re-used within tests, and vice versa.
This allows users to re-use skills acquired and strenghtens Puppet's better-together story.

**Alternatives:**
Bolt as communication layer touches most other components and - while having a well-defined interface - would not be easy to replace.
There is currently no reason to consider alternatives to Bolt.

## Configuration

Configuration for Litmus currently is spread over several files.

> The `provision.yml` contains various lists of platforms to target for tests.

Reviewer: ```suggestion
The `provision.yaml` contains various lists of platforms to target for tests.
```

The CI job setup (usually in `.travis.yml`, `appveyor.yml` or similar) contains the overall sequencing of steps for testing.
This setup is usually the same everywhere and is encoded in pdk-templates.

> There are sometimes slight variation to choose the puppet version and platform combinations from `provision.yml`.

Reviewer: ```suggestion
There are slight variations to choose the puppet version and platform combinations from `provision.yaml`.
```

The downside of virtual machines are the high resource usage and provisioning times in the order of minutes.
You can use local VMs on your development workstation, or private and public cloud providers.

To reduce resource usage and provisioning times docker and containers come in handy.
They deploy in seconds and achieve high packing densities with low overhead.
Best practices for container images justifiably frowns on SSH access or complete operating system services,
thus common public images are usually not representative of production on full VMs.
To avoid this limitation, [puppetlabs/litmusimages](https://github.com/puppetlabs/litmusimage) provides a set of pre-fabbed [docker images](https://hub.docker.com/u/litmusimage) that allow SSH access and have an init process running.

In some cases, using bare metal servers or already running systems is unavoidable.

### How to provision Test Infrastructure

There are as many ways to aqcuire access to test systems as there are kinds of test systems.
By default, Litmus calls out to a provisioner task to provision or tear down systems on demand.

> In the `puppetlabs-provision` module, we ship a number of commonly needed provisioners for docker (with SSH), docker_exp (using bolt's docker transport), vagrant (for local VMs) and the private cloud APIs [VMpooler](https://github.com/puppetlabs/vmpooler) and [ABS](https://github.com/puppetlabs/always-be-scheduling).

Reviewer: ABS is a private repository with a private tool designed for this provisioning, not sure if it's necessary to mention it

Author: ```suggestion
In the `puppetlabs-provision` module, we ship a number of commonly needed provisioners for docker (with SSH), docker_exp (using bolt's docker transport), vagrant (for local VMs) and the private cloud APIs [VMpooler](https://github.com/puppetlabs/vmpooler) and the puppet-private [ABS](https://github.com/puppetlabs/always-be-scheduling).
```

Author: I'll leave it in but mention that it is puppet-private.

Since Litmus 0.18 the rake tasks also allow calling arbitrary tasks outside the `provision` module to provision test systems. [Daniel's terraform demo](https://youtu.be/8BMo9DcZ4-Q) shows one application of this.

The provision task will allocate/provision/start VMs or containers of the requested platforms and add them to the inventory file.

> Through the use of the inventory file Litmus can also consume arbitrary other systems by users supplying their own inventory, independently of Litmus' provision capabilities.

Reviewer: ```suggestion
Through the use of the inventory file Litmus can also consume arbitrary other systems by users supplying their own data, independently of Litmus' provision capabilities.
```
